mypyc/genops.py

```diff
@@ -382,26 +382,56 @@ def __init__(self, fitem: FuncItem = INVALID_FUNC_DEF, name: str = '',
+        # that are nested inside of other functions.
+        self.callable_class = CallableClass()
         # Environment classes are ClassIR instances that contain attributes representing the
         # variables in the environment of the function they correspond to. Environment classes are
         # generated for functions that contain nested functions.
         self.env_class = INVALID_CLASS
-        # The register associated with the 'self' instance for function classes.
-        self.self_reg = INVALID_VALUE  # type: Value
-        # Environment class registers are the local registers associated with instances of an
-        # environment class, used for getting and setting attributes. env_reg is the register
-        # associated with the current environment, and prev_env_reg is the self.__mypyc_env__ field
-        # associated with the previous environment.
-        self.env_reg = INVALID_VALUE  # type: Value
-        self.prev_env_reg = INVALID_VALUE  # type: Value
         # These are flags denoting whether a given function is nested or contains a nested
         # function.
+        # Generator classes implement the '__next__' method, and are used to represent generators
+        # returned by generator functions.
+        self.generator_class = GeneratorClass()
+        # These are flags denoting whether a given function is nested, contains a nested function,
+        # or is a generator function.
         self.is_nested = False
         self.contains_nested = False
+        self.is_generator = fitem.is_generator
         # TODO: add field for ret_type: RType = none_rprimitive
+        self.curr_env_reg = INVALID_VALUE  # type: Value
+
+
+class ClassInfo(object):
+    def __init__(self) -> None:
+        # The ClassIR instance associated with this class.
+        self.ir = INVALID_CLASS
+        # The register associated with the 'self' instance for this generator class.
+        self.self_reg = INVALID_VALUE  # type: Value
+        # Environment class registers are the local registers associated with instances of an
+        # environment class, used for getting and setting attributes. curr_env_reg is the register
+        # associated with the current environment.
+        self.curr_env_reg = INVALID_VALUE  # type: Value
+
+
+class CallableClass(ClassInfo):
+    def __init__(self) -> None:
+        super().__init__()
+        # prev_env_reg is the self.__mypyc_env__ field associated with the previous environment.
+        self.prev_env_reg = INVALID_VALUE  # type: Value
+
+
+class GeneratorClass(ClassInfo):
+    def __init__(self) -> None:
+        super().__init__()
+        # This register holds the label number that the '__next__' function should go to the next
+        # time it is called.
+        self.next_label_reg = INVALID_VALUE  # type: Value
+        self.next_label_target = AssignmentTargetAttr(INVALID_VALUE, '')  # type: AssignmentTarget
```

Comment:
In general I don't love initializing stuff to `INVALID_VALUE` and the like. (We used to do a lot more of this, but I got rid of much of it.) How painful would it be to avoid using `INVALID_VALUE`? (By initializing them to useful values right away, using `Optional`, whatever.) It might be the least bad approach, but it is worth thinking on.

I find the `next_label_target` one particularly objectionable, I think.